### PR TITLE
protoc-gen-flowtypes: add required and nullable file and field extension

### PR DESCRIPTION
**Pull request by tmc**
protoc-gen-flowtypes: add required and nullable file and field extension

This enables closer expression of flow type system semantics

### CODE REVIEW

File: protoc-gen-flowtypes/opts/opts.proto [hunk 1]
```diff
@@ -0,0 +1,17 @@
+syntax = "proto2";
+
+import "google/protobuf/descriptor.proto";
+package opts;
+
+extend google.protobuf.FileOptions {
+  optional Options field_defaults = 1035;
```
Comment (johanbrandhorst) [inline protoc-gen-flowtypes/opts/opts.proto:7]: Is this number reserved? You can request a unique number from the protobuf team. I did it for protoc-gen-gopherjs.

Comment (johanbrandhorst) [reply protoc-gen-flowtypes/opts/opts.proto:7]: See https://github.com/google/protobuf/blob/9021f623e1420f513268a01a5ad43a23618a84ba/src/google/protobuf/descriptor.proto#L305

File: protoc-gen-flowtypes/genflowtypes/genflowtypes.go [hunk 2]
```diff
@@ -36,71 +37,78 @@ var knownTypeMap = map[string]string{
...
 }
 
 type objectFlowType struct {
-	Fields   []NamedFlowTyper
-	Options  Options
-	required bool
+	Fields  []NamedFlowTyper
+	Options Options
+
+	opts opts.Options
```
Comment (johanbrandhorst) [inline protoc-gen-flowtypes/genflowtypes/genflowtypes.go]: This struct has `Options` and `opts`, is there some way this confusing naming could be avoided? `FlowOptions`? I'm not familiar enough with the code but this stuck out to me as confusing.

File: protoc-gen-flowtypes/genflowtypes/genflowtypes.go [hunk 3]
```diff
@@ -210,11 +207,21 @@ func (cfg Options) messageToFlowType(m *descriptor.Message, reg *descriptor.Regi
...
 	for _, f := range m.Fields {
-		required := false
-		if validatorOptions := getFieldValidatorIfAny(f.FieldDescriptorProto); validatorOptions != nil {
-			required = *validatorOptions.MsgExists
+		/*
+			required := false
+			if validatorOptions := getFieldValidatorIfAny(f.FieldDescriptorProto); validatorOptions != nil {
+				required = *validatorOptions.MsgExists
+			}
+		*/
```
Comment (johanbrandhorst) [inline protoc-gen-flowtypes/genflowtypes/genflowtypes.go]: Remove comments?

**Review by johanbrandhorst — COMMENTED** [review]
In general I'm not sure this is the right change to make, since it goes back from proto3 to proto2 thinking. I read through https://github.com/google/protobuf/issues/1606#issuecomment-281832148 yesterday and largely agree that nullable and required fields probably cause more problems than they fix. I reserve judgement though since I'm not sure if this is particularly useful for flowtypes, the target of this generator.

File: protoc-gen-flowtypes/opts/opts.proto (hunk 1 above)
Comment (tmc) [reply protoc-gen-flowtypes/opts/opts.proto:7]: Yes this is reserved.

File: protoc-gen-flowtypes/genflowtypes/genflowtypes.go (hunk 2 above)
Comment (tmc) [reply protoc-gen-flowtypes/genflowtypes/genflowtypes.go]: Good point. I’ll consider a clarity pass.

**Review by alextoombs — APPROVED** [review]
LGTM aside from two options structs; think that should be changed pre-merge.

**Review by le-qi — APPROVED** [review]
lgtm

**Comment by tmc** [comment]
@johanbrandhorst I agree with you for the most part -- we're taking the position that these semantics are more appropriate for us for the time being.

**Pull request merged**